### ISSUE TITLE
2FA for logins

### ISSUE THREAD

**Issue by ara4n** [vector-im/riot-web#2772]
When I log in (using a username/password or 3PID/password combo), we should give users the option to also require a two factor authentication (or multi-factor authentication) via other channels.  Options are:

 * 2FA by email
 * 2FA by SMS (MSISDN)
 * 2FA by matrix, using another device.

**Comment by ara4n** [comment]
To do a sensible job of this we should probably have a seperate 3PID verification service, as proposed in https://github.com/matrix-org/synapse/issues/1710.

**Comment by Dasoren** [comment]
Is there any update on the work for this or anything?
Just wondering, because I would like 2FA for my account v.s just a username/password.

**Comment by spacekitteh** [comment]
Another crucial option would be via X.509, which would allow a large array of existing authentication methods to be used transparently.

**Comment by spacekitteh** [comment]
Relevant: https://github.com/matrix-org/olm/issues/5

**Comment by lpar** [comment]
[RFC6238](https://tools.ietf.org/html/rfc6238).

**Comment by vsund** [comment]
I'd really like to see [Fido U2F](https://en.wikipedia.org/wiki/Universal_2nd_Factor) like it's implemented by Yubico etc.

**Comment by lrvick** [comment]
I strongly econd U2F. It is well tested, standard, built to withstand phishing attacks, and never exposes any secrets to system memory.

**Comment by RyanSquared** [comment]
Correct me if I'm wrong but wouldn't this require it to be implemented in a Matrix implementation rather than Riot? Matrix would have to support U2F, with Riot just passing on the authentication request.

**Comment by t3chguy** [comment]
@RyanSquared yes: https://github.com/vector-im/riot-web/issues/2772#issuecomment-268310149

**Comment by lrvick** [comment]
@RyanSquared It would need to be implemented in a both the UI and the authentication server IIRC.

**Comment by RyanSquared** [comment]
So if it needs to be implemented on a backend server as well, is there a relevant issue for U2F or other 2FA methods on a backend server (or for the protocol)?

**Comment by singlerider** [comment]
+1 for U2F

**Comment by ara4n** [comment]
this is starting to get more urgent with the advent of cryptocommunities and other security-focused communities embracing Matrix. need to check out U2F and how it compares to TOTP and friends.

**Comment by pafcu** [comment]
FYI, Firefox does not support U2F out of the box currently, but it _was_ added to nightly last week, so should hopefully land at some point in the not too distant future.

**Comment by RyanSquared** [comment]
I'd say it's not "last week", based on this article: https://www.yubico.com/2017/09/firefox-nightly-enables-support-fido-u2f-security-keys/

**Comment by vsund** [comment]
Before Firefox 57 you can use an addon for Firefox to use U2F. I currently use U2F without any addons in Firefox 57 Beta. According to their release schedules, they release Firefox 57 on 2017-10-14 (https://wiki.mozilla.org/RapidRelease/Calendar). I had to enable it manually though.

**Comment by alduder** [comment]
U2F would be great, although it is so far supported by Firefox, Chrome and Opera only. A note on Firefox 57: Users must turn on the U2F switch (security.webauth.u2f) manually in the "about:config" settings.

**Comment by jaekwon** [comment]
We'll pay for this to be implemented by the end of January.  jae@tendermint.com

**Comment by tuxether** [comment]
Any updates on this?

**Comment by brainbug89** [comment]
Still no 2FA? 🙄

**Comment by jtl999** [comment]
In my mind, the authentication in Synapse should be easily "pluggable" so I could for example easily replace the standard username+password authentication with a custom plugin that takes a username+password+OTP as input. 

I'm unsure of how such a thing should be supported on the client side, maybe the server should be able to say "Need a username+password+OTP" from the client and Riot should add such fields to the login form?

**Comment by stevenaldinger** [comment]
bump just to see if vector im took up @jaekwon on offering to sponsor if nothing else.

2 factor isn't optional at this point imo but I really like the potential of this project so hate to bail on the thing. will contribute myself eventually but reinventing-the-wheel with my personal time is too much for a project I'm just now considering adopting.

my software engineering company would take the contract work to implement it for sure though 😄

**Comment by lrvick** [comment]
@stevenaldinger I for one would gladly vote with my wallet towards such an effort if the matrix team won't.

Can we get a bounty going for this somehow?

All the end to end encryption in the world is pointless if you can just phish someones login. This is still a major adoption blocker for any serious use cases.

**Comment by stevenaldinger** [comment]
let me scope it out a little bit and see what it really takes to make it a reality and I'll update. client is js which I'm really comfortable with and have worked with this sort of thing before, synapse/matrix server is python and I might make a mess lol (but with test coverage! 🤣), haven't looked into if they're both needing updates at this point or what the deal is. I'll try to lead the hunt this week though, I'm interested.

**Comment by ronmckown** [comment]
bump

**Comment by ronmckown** [comment]
Not having 2FA with fido(2) on a security platform is plainly rediculous.  At least throw in TOTP or even Google oauth.

**Comment by ptman** [comment]
An important thing to remember when implementing 2FA: allow for multiple second factors. IIRC AWS didn't. Allow any number (?) of U2F tokens and TOTP app tokens, preferably individually named. And recovery codes.

**Comment by jtl999** [comment]
> An important thing to remember when implementing 2FA: allow for multiple second factors. IIRC AWS didn't. Allow any number (?) of U2F tokens and TOTP app tokens, preferably individually named. And recovery codes.

Agreed. And the method of 2FA shouldn't be forced on HS operators, i.e I should be able to choose TOTP or U2F depending on my use case by using a certain plugin (if possible).

**Comment by sandys** [comment]
I would also like to add that this feature be combined with auth using Google Auth/Identity - https://cloud.google.com/identity/ since that is the most popular and compliant (with necessary certifications) auth platform in general.

**Comment by friedger** [comment]
While planning this feature, you should also take into account webauthn (https://www.w3.org/TR/webauthn/)

**Comment by RasmusKallas** [comment]
I reached out to yubico, they politely rejected investing time, sending me [this](https://developers.yubico.com/), what they already advertise on their web.

**Comment by Braintelligence** [comment]
I'm trying to decide on a communication platform and absolutely require 2fa. Is Matrix/Synapse/Riot capable of at least TOTP by now?
If not, how does the roadmap look after all those announced commitments of time and money currently?

**Comment by jtagcat** [comment]
> I'm trying to decide on a communication platform and absolutely require 2fa. Is Matrix/Synapse/Riot capable of at least TOTP by now?
> If not, how does the roadmap look after all those announced commitments of time and money currently?

Not yet implemented, though considering how fast security features are added, it's very likely that we will get multiple methods for 2fa. This is probably slower, since the feature does need to be supported on multiple platforms (client(s), server) and as a result in the matrix protocol.

**Comment by ptman** [comment]
One thing that is possible to do today is to use SAML and have your SAML IDP require 2FA

**Comment by Braintelligence** [comment]
@ptman So using Gluu with Matrix would be possible?

**Comment by dngray** [comment]
> While planning this feature, you should also take into account webauthn (https://www.w3.org/TR/webauthn/)

One of the things I like about WebAuthn is:

> The public key is not secret, because it is effectively useless without the corresponding private key. The fact that the server receives no secret has far-reaching implications for the security of users and organizations. Databases are no longer as attractive to hackers, because the public keys aren’t useful to them.

It seems [WebAuthn](https://en.wikipedia.org/wiki/WebAuthn) is a even more secure option than even [TOTP](https://en.wikipedia.org/wiki/Time-based_One-time_Password_algorithm) so I hope it doesn't overlooked just because everyone is already used to TOTP.

WebAuthn will have clear support in all browsers and that much is apparent being a part of the W3.

**Comment by RyanSquared** [comment]
> It seems WebAuthn is a even more secure option than even TOTP so I hope it doesn't overlooked just because everyone is already used to TOTP.

TOTP carries it's own set of issues with it; my Yubikey, for example, has a system similar to TOTP which I have sent to IRC channels quite often by mistake. I don't use it because of this reason.

It's less easy to phish a WebAuthn challenge given IIRC it's attached to the domain name of the source of authentication - however, I do recall an issue with web access to the USB interface where the challenge can be spoofed. Assuming the only way to access the "device" (as some systems such as Chromebooks and Apple hardware have it builtin) is through an interface provided by the web browser, it then becomes a lot harder, I believe impossible, situation to phish the WebAuthn "device" with a valid challenge.

**Comment by lpar** [comment]
It doesn't matter if a TOTP code gets disclosed publicly, unless someone knows what account it corresponds to, has cracked your password for that account, and is ready to try logging in during the appropriate 1 minute window.

958826 is the current TOTP code for my Google account. That doesn't even let you deduce what the next one is, even assuming you could work out exactly when a given key was generated.

**Comment by lrvick** [comment]
There seem to be some misconceptions in this thread that tend to pop up every time 2FA comes up on many tickets I track...

The point of 2FA is to make account compromise harder.

The most common way accounts are compromised are:

1. phishing
2. password cracking via database dumps
3. brute force cracking via a web interface

Working in security and infrastructure teams for most of my career I have seen a lot of #1, much less of #2, and almost never #3 anymore unless the target is a complete idiot that does not lock out an account for too many attempts from a given IP.

## Why SMS sucks

You can do a SIM Port attack or simply use a cheap SDR and a 2TB set of A/51 rainbow tables to dump codes out of the air in the US where carriers are forced to honor 2G retransmission requests.

It fails at reliably closing attack surfaces #1 and #2

## Why Yubico OTP sucks

Yubico-OTP is just AES challenge/response against a central Yubico-owned server which must share the secret, which I can tell you does -not- have 100% uptime and results in global account lockouts when it is down.

You can opt to host that server yourself, but then you end up with the AES secret exposed in the same application you are hoping does not get databse dumped.

Worse the Yubico OTP codes are -not- time based like TOTP and are in fact just a counter. This means you can capture one and use it hours later as long as a user has not logged in with a fresh one since.

It fails at #1 and #2 as well.

Yubikeys are a -solid- tool I document the use of extensively, but this is a feature that should be deprecated universally.

## Why TOTP sucks

TOTP fails at #1 because a phishing system prepared to automatically make immediate use of a password can simply do the same with a TOTP secret within a 30 second window without much fuss, and I have seen this happen in the wild.

TOTP normally fails at #2 because it relies on a shared secret that is combined with the current time, and the secret must be stored server side and be readily accessible in plain text. Unless someone happens to implement TOTP validation server-side in an HSM, which almost no one does, then an attacker with datbase access likely has access to the TOTP secrets as well.

TOTP is simply an awful design for 2FA that fails at the two most common forms of account compromise.

As an aside, Google Authenticator, the most popular implementation, just sores the TOTP secret in plain text in an sqlite database on both iOS and Android. I audited the code of both myself.

TOTP ALSO fails at #1 and #2 (Does no one actually do threat profiling before they design these stupid protocols?!)

## Why WebAuthn is awesome

It succeeds at threat #1 because the browser supplies the current domain name in your URL bar to your personal HSM (yubikey or otherwise) as part of the challenge/response flow. A malicious actor can't override this without the ability to manipulate core browser functionality not reachable by javascript (and if they can you have bigger problems).

It succeeds at #2 because the only thing the server needs to track is a public key, not unlike an authorized_keys file if you are familiar with ssh.

## Conclusion

WebAuthn is the only responsible form of 2FA worth implementing today, and we should be forcing users to use it on every service we can. I for one am working very hard to get SMS/TOTP/yubico-otp all deprecated and replaced with webauthn based solutions at every company I support. 

Even if someone does not have a personal HSM like a Yubikey there are software implementations available which still account for the most common forms of account compromise better than any alternatives like TOTP.

The lack of native WebAuthn support on Matrix is perhaps the biggest remaining black eye on the platform next to cross-signing which is already in progress.

Hopefully that argument was coherent but I am happy to answer any questions :)

**Comment by richvdh** [comment]
cf https://github.com/matrix-org/matrix-doc/pull/1998

**Comment by ricain** [comment]
@lrvick there a flaw in your use case: *password cracking via database dumps* should be spited in two:

1. password cracking from **your database dumps**
2. password cracking from **other website database dumps**

In case 1, you are totally right since shared secret for OTP is compromised.

In case 2, since your OTP shared secret is (by definition) not compromised, people who has a unique password for all theirs accounts are safe.

I agree that WebAuthn is better but cost are higher on setup.

**Comment by joepie91** [comment]
@irvick

> TOTP normally fails at #2 because it relies on a shared secret that is combined with the current time, and the secret must be stored server side and be readily accessible in plain text. Unless someone happens to implement TOTP validation server-side in an HSM, which almost no one does, then an attacker with datbase access likely has access to the TOTP secrets as well.

Nitpick: TOTP secrets are normally site-specific, no? So they don't affect the threat model of "someone cracks a password from your database to use on another site" - which is why we hash passwords, because your database has already been compromised *anyway* and so *for that* the attacker doesn't need to crack any passwords... the value is just in reusing the password elsewhere.

I do agree with your conclusion that WebAuthn is the way forward.

**Comment by RyanSquared** [comment]
> Nitpick: TOTP secrets are normally site-specific, no?

That should be correct. Section 3.R5 of RFC 6238 specifies that "there MUST be a unique secret (key) for each prover," where I believe a "prover" in this case is equivalent to a "site".

> So they don't affect the threat model of "someone cracks a password from your database to use on another site"

But that's not what this was referencing. This referenced the following section:

> an attacker with datbase access likely has access to the TOTP secrets as well.

lrvick's rule 2 states the following:

> password cracking via database dumps

This does not specify that it has to be a third party. I believe this was intended to be referencing the same database you'll be accessing with your next connection. This is solved by WebAuthn by having the secret key stored by only yourself rather than by the server.

If you have both a database leak of the system you're attacking, and know the password you're trying to use, it then becomes trivial. The only way to prevent against this, if you're using TOTP and are aware of the attack, is to then invalidate *every* TOTP secret and use some other kind of mechanism to reset them.

**Comment by esackbauer** [comment]
Well I am using the LDAP proxy of Duo.com for 2FA. Works great.

**Comment by caglar1** [comment]
has anyone tried to implement verifykit
https://verifykit.com/en/
it is free for 10.000 monthly verification
I think, it is the cheapest and convenient solution for 2fa

**Comment by esackbauer** [comment]
> has anyone tried to implement verifykit
> https://verifykit.com/en/
> it is free for 10.000 monthly verification
> I think, it is the cheapest and convenient solution for 2fa

Verification only via telegram or whatsapp... 
And only 1 app. 
Duo is free for up to 10 users, and you can have multiple apps protected. Verification via Duo app, which also replaces Google Authenticator.

**Comment by caglar1** [comment]
> > has anyone tried to implement verifykit
> > https://verifykit.com/en/
> > it is free for 10.000 monthly verification
> > I think, it is the cheapest and convenient solution for 2fa
> 
> Verification only via telegram or whatsapp...
> And only 1 app.
> Duo is free for up to 10 users, and you can have multiple apps protected. Verification via Duo app, which also replaces Google Authenticator.

Yes you right, But 10 free user is too limited,
There will be about 1000 users on my matrix server, 
so Duo would be so expensive for me

**Comment by ptman** [comment]
To repeat myself: Use synapse SAML support with a SAML IdP that can require 2FA, like Keycloak (free).

**Comment by thibaultmol** [comment]
Added a 30$ bounty on Bountysource for this feature request to be solved!
[![Bountysource](https://api.bountysource.com/badge/issue?issue_id=40250932)](https://www.bountysource.com/issues/40250932-feature-request-to-overlay-image-file?utm_medium=shield&utm_campaign=ISSUE_BADGE)

**Comment by epicfacethe3rd** [comment]
might give this a shot. commenting to bookmark it for myself

**Comment by rugk** [comment]
Wtf, Matrix, what happened to you? People would have paid you for this feature?

> We'll pay for this to be implemented by the end of January, just send us a service contract.

https://github.com/vector-im/riot-web/issues/2772#issuecomment-352947462

Also the [French goverment is a client of yours](https://matrix.org/blog/2018/04/26/matrix-and-riot-confirmed-as-the-basis-for-frances-secure-instant-messenger-app) that may very much appreciate such a feature…

**Comment by ptman** [comment]
> Also the [French goverment is a client of yours](https://matrix.org/blog/2018/04/26/matrix-and-riot-confirmed-as-the-basis-for-frances-secure-instant-messenger-app) that may very much appreciate such a feature…

They most likely use some kind of SSO (SAML? CAS? OpenID Connect?) that handles 2FA on another layer. Like keycloak. But I would not recommend SAML as you can now use OpenIDC.

**Comment by Agorise** [comment]
"2FA by matrix, using another device." - This is definitely needed, especially by a "Verified" alternate device (text, emoji, etc).

**Comment by saintlux897** [comment]
Pls I tried imputing my password for got hub but it anit going

**Comment by MarcusWichelmann** [comment]
Please keep in mind, that FIDO2/WebAuthn also supports password-less single factor authentication using only your hardware key (and maybe an additional PIN to unlock the device). Would be great, if you could support this too so passwords are no longer needed. The concept of passwords as an identification secret is fundamentally broken, anyway.

This would of course require mechanisms to register multiple authentication devices or generate backup keys so one doesn't lock itself out when loosing a device.

**Comment by jans23** [comment]
> Please keep in mind, that FIDO2/WebAuthn also supports password-less single factor authentication using only your hardware key (and maybe an additional PIN to unlock the device).

Once a PIN is required, it is two factor authentication.

**Comment by jtagcat** [comment]
jfyi It's the 'passwordless' movement you are defining.

In essence, it's just a more secure 1FA. The authentication merges to one factor during communication with the service.

Metaphorically speaking, if you have a key enclosed in a box, you open the box with a code to take out the key. You still use one key to open the door, as opposed to a keyhole and a pin on the door.

Nevertheless, PIN + hardware, considering most of the users and use cases, is still, likely to be more secure than just a passphrase/PIN.

**Comment by ara4n** [comment]
In case anyone is wondering why this hasn’t happened yet: we’ve found that most people who want 2FA are also using SSO, and so can use the SSO provider (keycloak etc) for this.

However, we still want to get it natively into Matrix, but it’s in the middle of the feature backlog.

**Comment by JonathanWilbur** [comment]
I'd like to point out that I have an [outstanding feature request](https://github.com/vector-im/element-web/issues/7967) that I reported for TLS / X.509 client certificate authentication. If implemented, it would require no change at all to Synapse, Dendrite, or the Matrix protocol, and would still provide an additional factor of very robust, well-understood authentication.

**Comment by networkException** [comment]
I believe a TLS certificate would not be what a regular user expects from a platform offering MFA. By looking at the comments in this issue it is clear that TOTP, U2F and FIDO2 / WebAuthn are prefered methods.

**Comment by Coolgam3r2152** [comment]
TOTP would be nice if added

**Comment by thalesfsp** [comment]
Since 2016.. and counting
